### DISCUSSION TITLE
AGW- When stopping magma@* services  or just mme -  S1 or sctp to enb stays as Connected

### DISCUSSION THREAD

**Discussion by pkapatsos-blinq** [magma/magma#6889] [Q&A]
Hi there, this is for AGW R1.4 and 1.5 , in earlier release this would shut down the sctp sessions.
![image](https://user-images.githubusercontent.com/61711855/118021647-7604f700-b329-11eb-955d-3a061427e5ce.png)
Think related to the stateless and or the sctpd being on its own ..Just sharing this is not good for enb vendors as if magma services are down. From the eNB/CPEs looks up but no data can pass. 
![image](https://user-images.githubusercontent.com/61711855/118022122-ffb4c480-b329-11eb-81a9-f9726390a76a.png)

Our ENB side with the above MME IP
![image](https://user-images.githubusercontent.com/61711855/118022455-62a65b80-b32a-11eb-94c9-d0e6d7429ca2.png)

just wondering if this is by design now or  only for unmanaged enb vendors etc .. 
We need the MME and SCTP to be synced  with other end or ENB S1 or our GUI and logs will be incorrect.
thanks.

**Comment by wallyrb** [comment]
Hey @pkapatsos-blinq thanks for sharing the information. Just to clarify sharing some information about stateless in AGW.

1. Restarting magma services would not kill S1. It takes about 30 seconds for the services to come back up and repopulate their last persisted state.
2. During restart UEs that are in ECM-CONNECTED state should not be impacted on the user plane side, i.e., their traffic should continue to flow.
3. For UEs in ECM-IDLE state, since the control plane communication will be down until all the services are up and running, they cannot move into ECM-CONNECTED state during restart.
4. For UEs that are not yet in EMM-REGISTERED state, they will not be able to attach to the network during service restart. 
5. After all the services are up, all the CP procedures should proceed as before.

Saying that, I think there are two problems here.

1. "From the eNB/CPEs looks up, but no data can pass". This is not expected, UE should be able to still browse with the magma service down until it recovers. Could you share the output of command `magmad_cli.py check_stateless`?.  If you can also share the mme and syslogs will be appreciated.
2. Discussing with @adityawasudeo, there was a service crash that originated this issue, could you provide more details or share if there is any open issue about this?

**Comment by pkapatsos-blinq** [comment]
@wallyrb thanks was just reverifying this the restart ok but just stopping the service, Steps 1-2 got it , I see so when we stop magma services manually which is the test case . Any UEs attached stay working and they do . Step 3-4  are the concern as the S1 does not show DOWN.  is there a timer that will restart the services is they are down ? 
This seems to be visual really as our eNB side shows S1 up which it is , however if magma is down sctpd is up so
Step5 - yes all back to normal.
now for the problem, 
1. this was our side our network related , Traffic does flow id UE is in ECM-CONNECTED State UE keeps working, why when we shut magma does the SCTPD stay up( it perhaps is 
root@magwr16-4:~# magmad_cli.py check_stateless
AGW Mode: STATELESS
 

2-NO service crash just negative test case , example behaviors to our HDW if magma services are down.


thanks

**Reply by wallyrb** [reply]
@pkapatsos-blinq  You can check in health.yml (under /etc/magma) there are two parameters restart_threshold and interval_check_mins

When condition of restart_threshold meets, recovery will trigger. Recovery means, sctpd will restart

**Reply by sudhikan** [reply]
Hey @pkapatsos-blinq - are you noticing that services do not gracefully come up after an induced restart and the outage (in the context of [3] and [4] above) is longer than 30 or so seconds?

**Comment by smiladi** [comment]
This issue was originally raised because we observed a scenario where UE and ENB think that UE is connected, but the UE does not exist on the AGW. Originally we thought this happened due to magma service restart, but we recently observed the issue again in our lab, and magma service was not restarted at that time.
What seems to cause the problem is that implicit detach timer expired on the AGW for number of UEs at the same time (around 25), and the ENB these UEs are connected to did not receive the UE Context Release for some of them.
It would be helpful if someone could describe a way to trigger explicit detach for all connected UEs, as that would make it easier to try to reproduce the issue; we will try to capture the exchange triggered by implicit detach, but that takes a lot of time and there is no guarantee the problem will re-appear.
The AGW is running 1.6. The issue was also observed  last week at customer site.

**Reply by joary** [reply]
Hi @smiladi,

> It would be helpful if someone could describe a way to trigger explicit detach for all connected UEs.

At the moment there's not a command to force the detach for a given subscriber. The only way of doing a bulk detach is by restarting both SCTPd and Magmad.

> What seems to cause the problem is that implicit detach timer expired on the AGW for number of UEs at the same time (around 25) and the ENB these UEs are connected to did not receive the UE Context Release for some of them.

Can you share the MME logs (/var/log/MME.info) for this scenario? Not clear from the description if this is scalability issue or a protocol bug for some UE config. Please include a pcap of the S1 interface as well.

In the scenario of this issue being a scalability issue, it will be harder to reproduce since this type of problems are bounded to the specific use-case and load. I'm looking for ideas on how to reproduce as well.

> but that takes a lot of time and there is no guarantee the problem will re-appear.

If you're willing to test, you can also reduce the [T3412 timer](https://github.com/magma/magma/blob/master/lte/gateway/configs/templates/mme.conf.template#L185) This will force the implicit detach to happen in 2*T3412 + 6 minutes.

**Reply by smiladi** [reply]
Hi @joary - please find attached. The CPEs went to IDLE around Thu Oct 07 23:03, then mobile reachability expired around Fri Oct 08 00:01 and finally implicit detach timer expired at Fri Oct 08 00:56. We'll try to reproduce it with shorter T3412 and debug level log.

[MME.pvmagw3.root.log.INFO.2021007-20211008.log.zip](https://github.com/magma/magma/files/7367473/MME.pvmagw3.root.log.INFO.2021007-20211008.log.zip)

**Reply by joary** [reply]
@smiladi, another two ask on this thread:

1. What is the inactivity timer configured in your RAN environment? 
    In the scenario we have been testing the eNB inactivity timer is fairly small 20s, which lead to eNB releasing the UE very early. In this scenario when the implicit detach takes place the eNB does not have anymore the UE state.
    
2. Can you share the syslogs for the same period above?
 Another possibility that we're looking at is AGW losing its state due to some Redis failure, will be awesome to double check that from syslogs.
   You can group all the necessary logs using the [show-tech tool](https://docs.magmacore.org/docs/lte/debug_show_tech#usage)

**Reply by joary** [reply]
> It would be helpful if someone could describe a way to trigger explicit detach for all connected UEs.

@smiladi, discussing this topic further we found out you can use the following command to detach a set of subscriber directly in the AGW:

```
s6a_service_cli.py delete <IMSI>,<IMSI>
```

**Reply by smiladi** [reply]
@joary - I tried the s6_service_cli command, but it doesn't seem it is working: 
- mobility_cli.py get_subscriber_table still shows these IMSIs and their IP addresses, 
- Wireshark capture doesn't show that AGW is sending any message to the ENB, and 
- UE remains connected

**Reply by joary** [reply]
@smiladi , a couple of triaging questions:

1. Which version of magma are you running (`apt policy magma`)? 
    I was able to see it working with `1.6.0-1625592603-1f26ba81`
    
1. Which command did you used exactly?
    Please observe the IMSI argument is just the number, for example to delete `IMSI001010000000001` you have to run `s6a_service_cli.py delete 001010000000001`

**Reply by smiladi** [reply]
@joary:
1. We are running the same version
2. The delete command with numbers does work (also, blanks as separators)
3. We repeated command number of times with 25 UEs and so far all UEs get deregistered every time. To be continued :)

**Reply by joary** [reply]
@smiladi , Just to clarify:

You're able to force the de-registration of the multiple subscribers at the same time, though you're not able to reproduce the problematic behavior described above?

**Reply by smiladi** [reply]
Hi @joary - that is correct. 
Though the scenario is not exactly the same as when the issue was detected, because when we execute delete command, the AGW thinks that UEs are still connected, and sends Detach Request first - see attached capture. 
[pi138-16.zip](https://github.com/magma/magma/files/7393667/pi138-16.zip)
Note that we also tried setting T3412 to 1 minute (and we can see in MME log that AGW really thinks that T3412 is 1 minute) and then waiting for 15 minutes, but even though there was no traffic during that time, the AGW did not move UEs to IDLE or sent UE Context Release to the ENB.

**Reply by joary** [reply]
@smiladi , 

> but even though there was no traffic during that time, the AGW did not move UEs to IDLE or sent UE Context Release to the ENB.

Even without traffic, if the UE is still reachable to the network after the first  T3412-time, it will send a Tracking-Area-Update to the core signalizing it is still alive. What you need to do is after the attach turn on airplane-mode in the UE (or simply power-off the UE). Then UE will not send TAU and later after 2*T3412+6min the UE should be detached from core.
